### ISSUE TITLE
[Feature Request] PMD suppress marker CLI option

### ISSUE THREAD

**Issue by bh-sergio-clara** [forcedotcom/sfdx-scanner#1184]
**Is your feature request related to a problem? Please describe.**
Having passed Security Review a few times we've developed our own comments/annotations to indicate "false positives". While these are mere guides for reviewers at the moment, since we add these to the lines identified by PMD, it would be useful to use our keywords as suppress markers. However this [CLI option](https://docs.pmd-code.org/latest/pmd_userdocs_suppressing_warnings.html) is not passed down to PMD so we'd have to "rewrite" and/or use "NOPMD" instead.

**Describe the solution you'd like**
Add the ability to use the `--suppress-marker` PMD argument.

**Additional context**
"Workaround": Use PMD directly but defeats the purpose of using the unified Salesforce Code Analyzer.

"Urgency": "Can't live without it". Considering the hundreds of annotated lines of code, we really must have this feature (or be "forced" to use PMD directly)/

**Comment by johnbelosf** [comment]
Thank you for raising this @bh-sergio-clara - I'll discuss with the team

**Comment by bh-shatz** [comment]
@bh-sergio-clara - although it is good to allow the `--suppress-marker` to be available when invoking the scanner, it will be even better to allow to configure individual rule with its own suppress marker to allow for specific comment format to only suppress specific type of rule.
I'm guessing this is actually a PMD requirement, not something for the sfdx-scanner.

**Comment by jfeingold35** [comment]
@bh-shatz , correct, this is something that would need to be implemented within PMD, not on our end. PMD's `// NOPMD` annotation suppresses all violations at the current line, while its `@SuppressWarnings()` annotation suppresses a single for the entire class. There's no way to suppress one rule at one line.

**Comment by stephen-carter-at-sf** [comment]
Just saw this issue... and believe we might be able to add to [Code Anaylzer v5](https://developer.salesforce.com/docs/platform/salesforce-code-analyzer/guide/code-analyzer.html)'s code-analyzer.yml configuration file something like:
```
engines:
  pmd:
    suppress_marker: "YOUR_CUSTOM_SUPPRESSION_MARKER_HERE"  # Default: NOPMD
```

But before doing so, I'm wondering if it is even needed for your use case. Do you want to omit a specific rule from running entirely? If so, if you use [Code Anaylzer v5](https://developer.salesforce.com/docs/platform/salesforce-code-analyzer/guide/code-analyzer.html) then you can remove the "Recommended" tag from any pmd rule that you do not want to run using our default `--rule-selector Recommended` from your code-analyzer.yml config file.

So this should serve your purposes even better than having to mark every single place the violation for that specific rule is found right?

@bh-shatz Please let us know.

**Comment by bh-sergio-clara** [comment]
@stephen-carter-at-sf We don't want do omit rules from running entirely, just looking to suppress specific rules on specific lines, instead of a wildcard suppression of all rules (like NOPMD). The idea being that just because something is, say a CRUD/FLS false positive it may still have SOQL injection issues that we don't want to inadvertently silence.

I think the configuration you're proposing could work but would need to be rule specific?

**Comment by stephen-carter-at-sf** [comment]
@bh-sergio-clara 
Does the PMD ```@SuppressWarnings("PMD.<RULE_NAME>")``` not work for your use case?
https://pmd.github.io/pmd/pmd_userdocs_suppressing_warnings.html

We have a code analyzer (non-pmd specific) solution that we have been brainstorming but it'll be at least 4 months before we can get to it... so please try out PMD's suppression mechanisms for now.

**Comment by bh-sergio-clara** [comment]
@stephen-carter-at-sf The issue with the `@SuppressWarnings` annotation is that is too broad making it easier to introduce issues in future. If there's a false positive of some sort that we want to suppress, with this annotation the whole function will be ignored (for that type of issue, but still).

The workaround we've managed so far involves a complex process of running PMD standalone, with different rules/markers combos and merging the results. Far from ideal but we can probably live with it for another 4 months. Thanks.

**Comment by stephen-carter-at-sf** [comment]
Thanks for your feedback. Yes the goal I'm going to be proposing is to be able to do line based violation suppression that works across all rules and engines in all file types that is Code Analyzer specific. But like I said, it is a few months down in our priority list so thanks for your patience in waiting.

**Comment by namrata111f** [comment]
@bh-sergio-clara We have released new inline marker suppression feature in Code Analyzer. Using this we can specify the markers around the particular lines of code where we want a violation to be suppressed. Please check the documentation at: https://developer.salesforce.com/docs/platform/salesforce-code-analyzer/guide/suppress-violations.html and let us know if further help is needed.
